config.org
#+title: PJ's GNU Emacs Config
#+author: Philip Johansson
#+description: PJ's personal Emacs config.
#+startup: showeverything
#+options: toc:2
* TABLE OF CONTENTS                                                     :toc:
- [[#important-initial-packages][Important initial packages]]
  - [[#use-package-package-manager][use-package package manager]]
  - [[#evil-mode][Evil mode]]
- [[#keybindings][Keybindings]]
  - [[#move-text][move-text]]
  - [[#escapekeyboard-escape-quit][escape=keyboard-escape-quit]]
  - [[#general][general]]
  - [[#keep-selection-on-indent][Keep selection on indent]]
  - [[#c-u][C-u]]
- [[#miscellaneous][Miscellaneous]]
  - [[#fonts][Fonts]]
  - [[#ui-tweaks][UI Tweaks]]
- [[#org-mode][Org mode]]
  - [[#enable-table-of-contents][Enable Table of Contents]]
  - [[#enable-org-bullets][Enable Org Bullets]]
  - [[#disable-electric-indent][Disable Electric Indent]]
- [[#sudo-edit][sudo-edit]]
- [[#rainbow-mode][rainbow mode]]
- [[#which-key][which-key]]
- [[#shells][Shells]]
  - [[#vterm][Vterm]]
- [[#theme][Theme]]
- [[#whitespace][Whitespace]]
- [[#git][git]]
  - [[#magit][magit]]
- [[#emacs-completion][emacs completion]]
- [[#multiple-cursors][multiple-cursors]]
- [[#code-completion][Code completion]]
  - [[#company][company]]
- [[#auto-save][Auto save]]
- [[#tabs-vs-space][Tabs vs Space]]
- [[#comment][Comment]]
- [[#dwim][DWIM]]
- [[#lsp][LSP]]
- [[#mu4e][mu4e]]
- [[#movement][Movement]]
- [[#treesitter][Treesitter]]

* Important initial packages
These are packages that are essential
** use-package package manager

#+begin_src emacs-lisp
  (setq package-enable-at-startup nil)
  (defvar bootstrap-version)
  (let ((bootstrap-file
          (expand-file-name
          "straight/repos/straight.el/bootstrap.el"
          (or (bound-and-true-p straight-base-dir)
              user-emacs-directory)))
          (bootstrap-version 7))
      (unless (file-exists-p bootstrap-file)
      (with-current-buffer
          (url-retrieve-synchronously
              "https://raw.githubusercontent.com/radian-software/straight.el/develop/install.el"
              'silent 'inhibit-cookies)
          (goto-char (point-max))
          (eval-print-last-sexp)))
      (load bootstrap-file nil 'nomessage))

  (straight-use-package 'use-package)

  (use-package diminish :straight t)
  (add-hook 'straight-use-package-mode-hook (lambda () (if (active-mode-p 'straight-use-package-mode) (straight-freeze-versions))))
#+end_src

** Evil mode
Very important, 'vim .gt. emacs'

#+begin_src emacs-lisp
  (defvar evil-want-keybinding nil)
  (defvar evil-want-C-u-scroll t)
  (defvar evil-want-integration t)

  (use-package evil
    :straight t
    :diminish
    :init
    (setq evil-vsplit-window-right t)
    (setq evil-split-window-below t)
    :config
    (evil-set-undo-system 'undo-redo)
    (evil-mode))

  (use-package evil-collection
    :straight t
    :diminish
    :after evil
    :config
    (evil-collection-init))
  (with-eval-after-load 'evil-collection
    (setq evil-collection-mode-list
          (remove 'vterm evil-collection-mode-list)))

  (with-eval-after-load 'evil
    (defalias #'forward-evil-word #'forward-evil-symbol)
    (setq-default evil-symbol-word-search t))
#+end_src
* Keybindings
** move-text
Move lines up and down with <C-j> and <C-k>

#+begin_src emacs-lisp
  (use-package move-text
    :straight t
    :config
    (evil-define-key 'normal 'global (kbd "C-k") #'move-text-up)
    (evil-define-key 'normal 'global (kbd "C-j") #'move-text-down))
#+end_src
** escape=keyboard-escape-quit
#+begin_src emacs-lisp
  (global-set-key (kbd "<escape>") 'keyboard-escape-quit)
#+end_src
** general
#+begin_src emacs-lisp
  (use-package general
    :straight t
    :config
    (general-evil-setup)
    (general-create-definer pj/leader-keys
      :states '(normal insert visual emacs)

      :keymaps 'override
      :prefix "SPC"
      :global-prefix "M-SPC")
    (pj/leader-keys
      "b" '(:ignore t :wk "buffer")
      "bb" '(switch-to-buffer :wk "Switch buffer")
      "bk" '(kill-current-buffer :wk "Kill buffer")
      "bn" '(next-buffer :wk "Buffer next")
      "bp" '(previous-buffer :wk "Buffer previous")
      "bq" '(bury-buffer :wk "Buffer bury")
      "br" '(revert-buffer :wk "Buffer reload"))

    (pj/leader-keys
      "c" '(:ignore t :wk "compile")
      "cc" '(compile :wk "Start compilation")
      "cn" '(next-error :wk "Next compilation error")
      "cp" '(previous-error :wk "Previous compilation error"))

    (pj/leader-keys
      "d" '(:ignore t :wk "dired")
      "dd" '(dired :wk "Open dired")
      "dD" '((lambda () (interactive) (dired "~")) :wk "Open dired in <HOME>"))

    (pj/leader-keys
      "g" '(:ignore: t :wk "magit")
      "gg" '(magit :wk "Open magit"))

    (pj/leader-keys
      "f" '(:ignore t :wk "file")
      "ff" '(find-file :wk "Find file")
      "fc" '((lambda () (interactive) (find-file "~/.config/emacs/config.org")) :wk "Open config.org")
      "fn" '(dired-create-empty-file :wk "New file"))

    (pj/leader-keys
      "e" '(:ignore t :wk "evaluate")
      "eb" '(eval-buffer :wk "Evaluate buffer")
      "ee" '(eval-expression :wk "Evaluate expression")
      "er" '(eval-region :wk "Evaluate region"))

    (pj/leader-keys
      "h" '(:ignore t :wk "help")
      "hf" '(describe-function :wk "Help function")
      "hv" '(describe-variable :wk "Help variable"))

    (pj/leader-keys
      "v" '(:ignore t :wk "vterm")
      "vv" '(multi-vterm :wk "Open vterm")
      "vn" '(multi-vterm-next :wk "Next vterm")
      "vp" '(multi-vterm-prev :wk "Previous vterm"))

    (pj/leader-keys
      "s" '(:ignore t :wk "split")
      "ss" '(split-window-right :wk "Split window vertically"))

    (pj/leader-keys
      "q"  '(evil-window-delete :wk "Quit window")
      "w"  '(save-buffer :wk "Save buffer")
      "/"  '(evil-commentary-line :wk "Comment line/region"))

    (pj/leader-keys
      "m" '(:ignore t :wk "multiple-cursors")
      "mm" '(mc/edit-lines :wk "Mark at selection")
      "mn" '(mc/mark-next-like-this :wk "Mark next like this")
      "mp" '(mc/mark-previous-like-this :wk "Mark prev like this")
      "ma" '(mc/mark-all-like-this :wk "Mark all like this"))
    )
#+end_src
** Keep selection on indent
If text is selected when '<' or '>', keep the selection

#+begin_src emacs-lisp
  (define-key evil-visual-state-map (kbd "<") (lambda ()
                                                (interactive)
                                                (call-interactively 'evil-shift-left)
                                                (evil-normal-state)
                                                (evil-visual-restore)))
  (define-key evil-visual-state-map (kbd ">") (lambda ()
                                                (interactive)
                                                (call-interactively 'evil-shift-right)
                                                (evil-normal-state)
                                                (evil-visual-restore)))
#+end_src
** C-u
For some reason evil can't handle this one
#+begin_src emacs-lisp
  (define-key evil-motion-state-map (kbd "C-u") 'evil-scroll-up)
#+end_src
* Miscellaneous
** Fonts
What fonts to use, very important

#+begin_src emacs-lisp
  (set-frame-font "Fira Code 13")

  (setq-default line-spacing 0.12)
#+end_src

*** Zooming In/Out
You can use the bindings CTRL plus =/- for zooming in/out.
You can also use CTRL plus the mouse wheel for zooming in/out.

#+begin_src emacs-lisp
  (global-set-key (kbd "C-=") 'text-scale-increase)
  (global-set-key (kbd "C--") 'text-scale-decrease)
  (global-set-key (kbd "<C-wheel-up>") 'text-scale-increase)
  (global-set-key (kbd "<C-wheel-down>") 'text-scale-decrease)
#+end_src
*** grep
#+begin_src emacs-lisp
  (setq-default grep-command "rg --vimgrep")
#+end_src
** UI Tweaks
#+begin_src emacs-lisp
  (setq inhibit-startup-screen t)      ;; no splash screen
  (menu-bar-mode 0)                    ;; ew menu bar
  (tool-bar-mode 0)                    ;; ew tool bar
  (scroll-bar-mode 0)                  ;; ew scroll bar
  (column-number-mode 1)               ;; so we can see the column
  (global-display-line-numbers-mode 1) ;; line numbers are useful
  (show-paren-mode 1)                  ;; on by default but still
  (setq show-trailing-whitespace t)
  (fset 'yes-or-no-p 'y-or-n-p)
  (display-time)
  (global-hl-line-mode 1)
  (add-hook 'compilation-filter-hook 'ansi-color-compilation-filter) ;; For ANSI colors
  (add-hook 'comint-output-filter-functions 'ansi-color-process-output)
#+end_src
* Org mode
** Enable Table of Contents

#+begin_src emacs-lisp
  (use-package toc-org
    :straight t
    :commands toc-org-enable
    :init (add-hook 'org-mode-hook 'toc-org-enable))
#+end_src
** Enable Org Bullets
Org-bullets gives us attractive bullets rather than asterisks.

#+begin_src emacs-lisp
  (add-hook 'org-mode-hook 'org-indent-mode)
  (use-package org-bullets :straight t)
  (add-hook 'org-mode-hook (lambda ()
                             (org-bullets-mode 1)
                             (diminish 'org-indent-mode)))
#+end_src
** Disable Electric Indent
Org mode source blocks have some really weird and annoying default indentation behavior.
I think this has to do with electric-indent-mode, which is turned on by default in Emacs. So let's turn it OFF!

#+begin_src emacs-lisp
  (electric-indent-mode -1)
#+end_src

#+begin_src emacs-lisp
  (defun my/org-reformat-buffer ()
    "Reformat an org-mode buffer."
    (interactive)
    (when (y-or-n-p "Format current buffer? ")
      (let ((document (org-element-interpret-data (org-element-parse-buffer))))
        (erase-buffer)
        (insert document)))
    (message nil))
#+end_src

*** Colors
#+begin_src emacs-lisp
  (set-face-attribute 'org-meta-line nil :height 0.8 :slant 'normal :foreground "#555555")
  (custom-set-faces '(org-block ((t (:background "#181818" :foreground "#e4e4ef")))))
#+end_src
* sudo-edit
#+begin_src emacs-lisp
  (use-package sudo-edit
    :straight t
    :config
    (pj/leader-keys
      "fu" '(sudo-edit-find-file :wk "Sudo find file")
      "fU" '(sudo-edit :wk "Sudo edit file")))
#+end_src
* rainbow mode
Display the actual color as a background for any hex color value (ex. #abcdef).
The code block below enables rainbow-mode in all programming modes (prog-mode) as well as org-mode,
which is why rainbow works in this document.

#+begin_src emacs-lisp
  (use-package rainbow-mode
    :straight t
    :diminish
    :hook
    ((org-mode prog-mode) . rainbow-mode))
#+end_src
* which-key
#+begin_src emacs-lisp
  (use-package which-key
    :straight t
    :diminish
    :init
    (setq which-key-idle-delay 0.2)
    :config
    (setq  which-key-side-window-location 'bottom
           which-key-sort-order #'which-key-key-order-alpha
           which-key-add-column-padding 1
           which-key-sort-uppercase-first nil
           which-key-max-display-columns nil
           which-key-min-display-lines 10
           which-key-side-window-slot +10
           which-key-side-window-max-height 0.40
           which-key-max-description-length 25
           which-key-allow-imprecise-window-fit t)
    (which-key-mode 1))
#+end_src
* Shells
** Vterm
Vterm is a terminal emulator within Emacs.
The 'shell-file-name' setting sets the shell to be used in M-x shell, M-x term, M-x ansi-term and M-x vterm.
By default, the shell is set to 'fish' but could change it to 'bash' or 'zsh' if you prefer.

#+begin_src emacs-lisp
  (setq shell-file-name "/usr/bin/fish")

  (use-package vterm
    :straight t
    :config
    (setq vterm-max-scrollback 5000))
  (use-package multi-vterm
    :straight t)
#+end_src
* Theme
Less is more

#+begin_src emacs-lisp
  (use-package gruber-darker-theme
    :straight t)
  (custom-set-faces
    '(org-block ((t (:background "#181818" :foreground "#e4e4ef")))))
#+end_src
* Whitespace
So we can see that damn whitespace

#+begin_src emacs-lisp
  (use-package whitespace
    :straight t
    :diminish
    :hook ((org-mode prog-mode) . whitespace-mode)
    :config
    (setq whitespace-style '(face tabs spaces trailing space-before-tab indentation space-after-tab space-mark tab-mark missing-newline-at-eof)))
#+end_src
* git
** magit
Duh

#+begin_src emacs-lisp
  (use-package magit
    :straight t
    :diminish)

  (setq magit-diff-refine-hunk (quote all))
#+end_src

#+begin_src emacs-lisp
  (use-package diff-hl
    :straight t
    :diminish
    :config
    (global-diff-hl-mode)
    (custom-set-faces
     '(diff-hl-change ((t (:background "#333355" ))))
     '(diff-hl-delete ((t (:background "#553333" ))))
     '(diff-hl-insert ((t (:background "#335533" ))))))
  (setq diff-hl-show-hunk-inline-popup-smart-lines nil)
#+end_src
* emacs completion
#+begin_src emacs-lisp
  (use-package vertico
    :straight t
    :config (vertico-mode))
  (use-package orderless
    :straight t
    :custom
    (completion-styles '(orderless basic)))

  (use-package orderless
    :straight t
    :custom
    (completion-styles '(flex))
    (completion-category-defaults nil))

  (use-package marginalia
    :straight t
    :config
    (marginalia-mode))
#+end_src
* multiple-cursors
#+begin_src emacs-lisp
  (use-package multiple-cursors
    :straight t)
#+end_src
* Code completion
** company
#+begin_src emacs-lisp
  (use-package company
    :straight t
    :diminish)
  (add-hook 'after-init-hook 'global-company-mode)
#+end_src
* Auto save
#+begin_src emacs-lisp
  (setq backup-directory-alist '(("." . "~/.cache/emacs")))
#+end_src
* Tabs vs Space
#+begin_src emacs-lisp
  (setq-default indent-tabs-mode nil)
  (setq-default tab-width 4)
  (use-package dtrt-indent
    :straight t
    :config
    (setq dtrt-indent-global-mode 1))
  (add-hook 'dts-mode-hook (lambda () (setq indent-tabs-mode t) (setq tab-width 4)))
#+end_src
* Comment
#+begin_src emacs-lisp
  (use-package evil-commentary
    :diminish
  :straight t
  :config
  (evil-commentary-mode))
#+end_src
* DWIM
#+begin_src emacs-lisp
  (setq dired-dwim-target t)
#+end_src
* LSP
#+begin_src emacs-lisp
  ;; (setq eldoc-echo-area-use-multiline-p nil)
  ;; (setq eldoc-echo-area-prefer-doc-buffer t)
  ;; (evil-define-key 'normal 'global (kbd "S-k") 'eldoc)
  ;; (add-hook 'eglot-managed-mode-hook (lambda () (eglot-inlay-hints-mode t)))
  ;; (diminish 'eldoc-mode)
  ;; (diminish 'global-eldoc-mode)
  ;; (diminish 'flymake-mode)

  ;; (add-hook 'rust-mode #'eglot-ensure)
  ;; (add-hook 'c-mode #'eglot-ensure)
#+end_src
* mu4e
Not working
#+begin_src emacs-lisp
  (add-to-list 'load-path "~/.local/share/emacs/site-lisp/mu4e")
#+end_src
* Movement
#+begin_src emacs-lisp
  (define-key global-map (kbd "M-l") 'evil-window-right)
  (define-key global-map (kbd "M-h") 'evil-window-left)
  (define-key global-map (kbd "M-k") 'evil-window-up)
  (define-key global-map (kbd "M-j") 'evil-window-down)
#+end_src
* Treesitter
#+begin_src emacs-lisp
  (setq treesit-font-lock-level 4)
  (setq treesit-language-source-alist
     '((bash "https://github.com/tree-sitter/tree-sitter-bash")
       (c "https://github.com/tree-sitter/tree-sitter-c" "v0.20.0")
       ))
#+end_src


config.el
(require 'package)
(setq package-archives
  '(("melpa" . "https://melpa.org/packages/")
    ("gnu" . "https://elpa.gnu.org/packages/")
    ("nongnu" . "https://elpa.nongnu.org/nongnu/")))
(package-initialize)

(unless (package-installed-p 'use-package)
(package-refresh-contents)
(package-install 'use-package))

(require 'use-package)
(use-package diminish :ensure t)

(defvar evil-want-keybinding nil)
  (defvar evil-want-C-u-scroll t)
  (defvar evil-want-integration t)

  (use-package evil
    :init
    (setq evil-vsplit-window-right t)
    (setq evil-split-window-below t)
    :config
    (evil-set-undo-system 'undo-redo)
    (evil-mode))

(use-package evil-collection
  :ensure t
  :after evil
  :config
  (evil-collection-init))

(use-package general
  :ensure t
  :config
  (general-evil-setup)
  (general-create-definer pj/leader-keys
    :states '(normal insert visual emacs)

    :keymaps 'override
    :prefix "SPC"
    :global-prefix "M-SPC")
  (pj/leader-keys
    "b" '(:ignore t :wk "buffer")
    "bb" '(switch-to-buffer :wk "Switch buffer")
    "bk" '(kill-this-buffer :wk "Kill buffer")
    "bn" '(next-buffer :wk "Buffer next")
    "bp" '(previous-buffer :wk "Buffer previous")
    "br" '(revert-buffer :wk "Buffer reload"))
  
  (pj/leader-keys
    "cc" '(:ignore t :wk "compile")
    "cn" '(next-error :wk "Next compilation error")
    "cp" '(previous-error :wk "Previous compilation error"))

  (pj/leader-keys
    "d" '(:ignore t :wk "dired")
    "dd" '(dired :wk "Open dired"))

  (pj/leader-keys
    "g" '(:ignore: t :wk "magit")
    "gg" '(magit :wk "Open magit"))

  (pj/leader-keys
    "f" '(:ignore t :wk "file")
    "ff" '(find-file :wk "Find file")
    "fc" '((lambda () (interactive) (find-file "~/.config/emacs/config.org")) :wk "Open config.org"))

  (pj/leader-keys
    "e" '(:ignore t :wk "evaluate")
    "eb" '(eval-buffer :wk "Evaluate buffer")
    "ed" '(eval-defun :wk "Evaluate defun")
    "ee" '(eval-expression :wk "Evaluate expression")
    "er" '(eval-region :wk "Evaluate region"))

  (pj/leader-keys
    "h" '(:ignore t :wk "help")
    "hf" '(describe-function :wk "Help function")
    "hv" '(describe-variable :wk "Help variable"))

  (pj/leader-keys
    "v" '(:ignore t :wk "vterm")
    "vv" '(multi-vterm :wk "Open vterm")
    "vn" '(multi-vterm-next :wk "Next vterm")
    "vp" '(multi-vterm-prev :wk "Previous vterm"))

  (pj/leader-keys
    "s" '(:ignore t :wk "split")
    "ss" '(split-window-right :wk "Split window vertically"))

  (pj/leader-keys
    "q"  '(evil-window-delete :wk "Quit window")
    "w"  '(save-buffer :wk "Save buffer")
    "/"  '(comment-line :wk "Comment line/region"))

  (pj/leader-keys
    "m" '(:ignore t :wk "multiple-cursors")
    "mm" '(mc/edit-lines :wk "Mark at selection")
    "mn" '(mc/mark-next-like-this :wk "Mark next like this")
    "mp" '(mc/mark-previous-like-this :wk "Mark prev like this")
    "ma" '(mc/mark-all-like-this :wk "Mark all like this"))
)

(use-package move-text
  :ensure t
  :config
  (evil-define-key 'normal 'global (kbd "C-j") #'move-text-line-down)
  (evil-define-key 'normal 'global (kbd "C-k") #'move-text-line-up))

(define-key evil-visual-state-map (kbd "<") (lambda ()
                                              (interactive)
                                              (call-interactively 'evil-shift-left)
                                              (evil-normal-state)
                                              (evil-visual-restore)))
(define-key evil-visual-state-map (kbd ">") (lambda ()
                                              (interactive)
                                              (call-interactively 'evil-shift-right)
                                              (evil-normal-state)
                                              (evil-visual-restore)))

(define-key evil-motion-state-map (kbd "C-u") 'evil-scroll-up)

(set-face-attribute 'default nil
  :font "FiraCode Nerd Font"
  :height 180)

;; Uncomment the following line if line spacing needs adjusting.
(setq-default line-spacing 0.12)

(global-set-key (kbd "C-=") 'text-scale-increase)
(global-set-key (kbd "C--") 'text-scale-decrease)
(global-set-key (kbd "<C-wheel-up>") 'text-scale-increase)
(global-set-key (kbd "<C-wheel-down>") 'text-scale-decrease)

(setq inhibit-startup-screen t)      ;; no splash screen
(menu-bar-mode 0)                    ;; ew menu bar
(tool-bar-mode 0)                    ;; ew tool bar
(scroll-bar-mode 0)                  ;; ew scroll bar
(column-number-mode 1)               ;; so we can see the column
(global-display-line-numbers-mode 1) ;; line numbers are useful
(show-paren-mode 1)                  ;; on by default but still

(use-package toc-org
  :ensure t
  :commands toc-org-enable
  :init (add-hook 'org-mode-hook 'toc-org-enable))

(add-hook 'org-mode-hook 'org-indent-mode)
(use-package org-bullets :ensure t)
(add-hook 'org-mode-hook (lambda () (org-bullets-mode 1)))

(electric-indent-mode -1)

(use-package sudo-edit
  :ensure t
  :config
    (pj/leader-keys
      "fu" '(sudo-edit-find-file :wk "Sudo find file")
      "fU" '(sudo-edit :wk "Sudo edit file")))

(use-package rainbow-mode
  :ensure t
  :diminish
  :hook 
  ((org-mode prog-mode) . rainbow-mode))

(use-package which-key
  :ensure t
  :init
    (which-key-mode 1)
  :config
  (setq which-key-side-window-location 'right
	  which-key-sort-order #'which-key-key-order-alpha
	  which-key-sort-uppercase-first nil
	  which-key-add-column-padding 1
	  which-key-max-display-columns nil
	  which-key-min-display-lines 6
	  which-key-side-window-slot -10
	  which-key-side-window-max-height 0.25
	  which-key-idle-delay 0.2
	  which-key-max-description-length 40
	  which-key-allow-imprecise-window-fit t
	  which-key-separator " → " ))

(use-package vterm
  :ensure t
  :config
  (setq shell-file-name "/bin/fish"
        vterm-max-scrollback 5000))
(use-package multi-vterm
  :ensure t)

(use-package gruber-darker-theme
  :ensure t)
(custom-set-faces
  '(org-block ((t (:background "#181818" :foreground "#e4e4ef")))))

(use-package whitespace
  :ensure t
  :diminish
  :hook ((org-mode prog-mode) . whitespace-mode)
  :config
  (setq whitespace-style '(face spaces space-mark tab-mark tabs))
  (setq whitespace-space-regexp "\\(\x20\\)"))

(use-package magit
:ensure t
:diminish)

(use-package ido
  :ensure t
  :config
  (ido-mode 1)
  (ido-everywhere 1)
  (setq ido-enable-flex-matching t)
  (setq-default confirm-nonexistent-file-or-buffer nil)
  (setq ido-use-filename-at-point 'guess))

(use-package ido-completing-read+
  :ensure t
  :after ido
  :config
  (ido-ubiquitous-mode 1))

(use-package ido-vertical-mode
  :ensure t
  :config
  (ido-vertical-mode 1)
  (setq ido-vertical-define-keys 'C-n-and-C-p-only))

(use-package multiple-cursors
  :ensure t)
